python/GafferUI/PlugValueWidget.py

```diff
@@ -399,80 +457,88 @@ def _popupMenuDefinition( self ) :
 
 	def __nodeMetadataChanged( self, nodeTypeId, key, node ) :
 
-		if self.__plug is None :
-			return
-
-		if Gaffer.MetadataAlgo.readOnlyAffectedByChange( self.__plug, nodeTypeId, key, node ) :
-			self._updateFromPlug()
+		for p in self.__plugs :
+			if Gaffer.MetadataAlgo.readOnlyAffectedByChange( p, nodeTypeId, key, node ) :
+				self._updateFromPlugs()
+				return
 
 	def __contextChanged( self, context, key ) :
 
-		self._updateFromPlug()
+		self._updateFromPlugs()
+
+	def __setPlugsInternal( self, plugs, callUpdateFromPlugs ) :
+
+		assert( isinstance( plugs, set ) )
+		if len( plugs ) and sole( p.__class__ for p in plugs ) is None :
+			raise ValueError( "Plugs have different types" )
 
-	def __setPlugInternal( self, plug, callUpdateFromPlug ) :
+		nodes = set()
+		scriptNodes = set()
+		for plug in plugs :
+			nodes.add( plug.node() )
+			scriptNodes.add( plug.ancestor( Gaffer.ScriptNode ) )
 
-		self.__plug = plug
+		assert( len( scriptNodes ) <= 1 )
 
-		context = self.__fallbackContext
+		self.__plugs = plugs
 
-		if self.__plug is not None :
-			self.__plugDirtiedConnection = plug.node().plugDirtiedSignal().connect( Gaffer.WeakMethod( self.__plugDirtied ) )
-			self.__plugInputChangedConnection = plug.node().plugInputChangedSignal().connect( Gaffer.WeakMethod( self.__plugInputChanged ) )
+		self.__plugDirtiedConnections = [
+			node.plugDirtiedSignal().connect( Gaffer.WeakMethod( self.__plugDirtied ) )
+			for node in nodes
+		]
+		self.__plugInputChangedConnections = [
+			node.plugInputChangedSignal().connect( Gaffer.WeakMethod( self.__plugInputChanged ) )
+			for node in nodes
+		]
+
+		if self.__plugs :
 			self.__plugMetadataChangedConnection = Gaffer.Metadata.plugValueChangedSignal().connect( Gaffer.WeakMethod( self.__plugMetadataChanged ) )
-			scriptNode = self.__plug.ancestor( Gaffer.ScriptNode )
-			if scriptNode is not None :
-				context = scriptNode.context()
 		else :
-			self.__plugDirtiedConnection = None
-			self.__plugInputChangedConnection = None
 			self.__plugMetadataChangedConnection = None
 
-		self.__context = context
+		scriptNode = next( iter( scriptNodes ), None )
+		self.__context = scriptNode.context() if scriptNode is not None else self.__fallbackContext
 		self.__updateContextConnection()
 
-		if callUpdateFromPlug :
-			self._updateFromPlug()
+		if callUpdateFromPlugs :
+			self._updateFromPlugs()
 
 	def __updateContextConnection( self ) :
 
-		# we only want to be notified of context changes if we have a plug and that
-		# plug has an incoming connection. otherwise context changes are irrelevant
-		# and we'd just be slowing things down by asking for notifications.
+		# We only want to be notified of context changes for plugs whose values are
+		# computed.
 
 		context = self.__context
-		plug = self.getPlug()
-		if plug is None or ( plug.getInput() is None and plug.direction() == Gaffer.Plug.Direction.In ):
+		if all( p.source().direction() == Gaffer.Plug.Direction.In for p in self.getPlugs() ) :
```

Comment:
We have, but intentionally. What we're trying to detect is situations where the plug value isn't the result of a compute, meaning we don't need to track context changes. We could have an input, but if it too is an input plug rather than an output, it's value won't be computed either. That's why we now compare `source().direction()`. It's a stronger check that will result in less unnecessary signal handling.

Reply:
Yeah thats way better. Sorry i'd missed the fact it was now `p.source().direction()`.